safehttp/plugins/xsrf/xsrf.go

```diff
@@ -18,146 +18,229 @@
-
-// Token extracts the XSRF token from the incoming request. If it is not
-// present, it returns a non-nil error.
-func Token(r *safehttp.IncomingRequest) (string, error) {
-	tok := r.Context().Value(tokenCtxKey{})
-	if tok == nil {
-		return "", errors.New("xsrf token not found")
+func New(key string, c Checker, i Injector) Interceptor {
+	return Interceptor{
+		secretAppKey: key,
+		c:            c,
+		i:            i,
 	}
-	return tok.(string), nil
 }
 
-func addCookieID(w *safehttp.ResponseWriter) (*safehttp.Cookie, error) {
-	buf := make([]byte, 20)
-	if _, err := rand.Read(buf); err != nil {
-		return nil, fmt.Errorf("crypto/rand.Read: %v", err)
+func Default(key string) Interceptor {
+	return Interceptor{
+		secretAppKey: key,
+		c: defaultChecker{
+			secretAppKey: key,
+			cookieIDKey:  "xsrf-cookie",
+			tokenKey:     "xsrf-token",
+		},
+		i: defaultInjector{
+			secretAppKey: key,
+			cookieIDKey:  "xsrf-cookie",
+		},
 	}
+}
 
-	c := safehttp.NewCookie(cookieIDKey, base64.StdEncoding.EncodeToString(buf))
-	c.SetSameSite(safehttp.SameSiteStrictMode)
-	if err := w.SetCookie(c); err != nil {
-		return nil, err
+func Angular(cookieName, headerName string) Interceptor {
+	return Interceptor{
+		c: angularChecker{
+			tokenCookieName: cookieName,
+			tokenHeaderName: headerName,
+		},
+		i: angularInjector{
+			tokenCookieName: cookieName,
+		},
 	}
-	return c, nil
 }
 
-// Before should be executed before directing the safehttp.IncomingRequest to
-// the handler to ensure it is not part of a Cross-Site Request
-// Forgery attack.
-//
-// On first user visit through a state preserving request (GET, HEAD or
-// OPTIONS), a nonce-based cookie will be set in the response as a way to
-// distinguish between users and prevent pre-login XSRF attacks. The cookie will
-// be used in the token generation and verification algorithm and is expected to
-// be present in all subsequent incoming requests.
-//
-// For every authorized request, the interceptor will also generate a
-// cryptographically-safe XSRF token using the appKey, the cookie and the path
-// visited. This can be later extracted using Token and should be injected as a
-// hidden input field in HTML forms.
-//
-// In case of state changing requests (all except GET, HEAD and OPTIONS), the
-// interceptor checks for the presence of the XSRF token in the request body
-// (expected to have been injected) and validates it.
 func (it *Interceptor) Before(w *safehttp.ResponseWriter, r *safehttp.IncomingRequest, _ safehttp.InterceptorConfig) safehttp.Result {
-	needsValidation := !statePreservingMethods[r.Method()]
-	cookieID, err := r.Cookie(cookieIDKey)
+	if statePreservingMethods[r.Method()] {
+		return safehttp.NotWritten()
+	}
+	tok, userID, actionID, code := it.c.Retrieve(r)
+	if code != safehttp.StatusOK {
+		return w.WriteError(code)
+	}
+	code = it.c.Validate(tok, userID, actionID)
+	if code != safehttp.StatusOK {
+		return w.WriteError(code)
+	}
+	return safehttp.NotWritten()
```

Comment:
@kele I think we might have overdone the abstraction bit a tad too much.

All of this CL and all of this work is just to allow us to not repeat these 12 lines?

Reply:
> @kele I think we might have overdone the abstraction bit a tad too much.

Hard for me to say until I see the end result. Right now it is still super weird that `Retrieve` and `Validate` belong to the same interface. `Retrieve` can be front-end specific, whereas `Validate` can be crypto-specific. And if we think the second argument is not possible enough, then `Retrieve` and `Validate` should simply be one call: `Check()` and then it all gets simplified.